plugins/vpp/binapi/vpp1908/punt/punt.ba.go

```diff
@@ -22,58 +25,290 @@ var _ = bytes.NewBuffer
+	return "ip6_prefix"
+}
+func (*IP6Prefix) GetCrcString() string {
+	return "779fd64f"
+}
+
+// Mprefix represents VPP binary API type 'mprefix':
+type Mprefix struct {
+	Af               AddressFamily
+	GrpAddressLength uint16
+	GrpAddress       AddressUnion
+	SrcAddress       AddressUnion
+}
+
+func (*Mprefix) GetTypeName() string {
+	return "mprefix"
+}
+func (*Mprefix) GetCrcString() string {
+	return "1c4cba05"
+}
+
+// Prefix represents VPP binary API type 'prefix':
+type Prefix struct {
+	Address       Address
+	AddressLength uint8
+}
+
+func (*Prefix) GetTypeName() string {
+	return "prefix"
+}
+func (*Prefix) GetCrcString() string {
+	return "0403aebc"
+}
+
 // Punt represents VPP binary API type 'punt':
 type Punt struct {
-	IPv        uint8
-	L4Protocol uint8
-	L4Port     uint16
+	Type PuntType
+	Punt PuntUnion
 }
 
 func (*Punt) GetTypeName() string {
 	return "punt"
 }
 func (*Punt) GetCrcString() string {
-	return "fe4f98ac"
+	return "b9885aa7"
+}
+
+// PuntException represents VPP binary API type 'punt_exception':
+type PuntException struct {
+	ID uint32
+}
+
+func (*PuntException) GetTypeName() string {
+	return "punt_exception"
+}
+func (*PuntException) GetCrcString() string {
+	return "da11ca36"
+}
+
+// PuntL4 represents VPP binary API type 'punt_l4':
+type PuntL4 struct {
+	Af       AddressFamily
+	Protocol IPProto
+	Port     uint16
+}
+
+func (*PuntL4) GetTypeName() string {
+	return "punt_l4"
+}
+func (*PuntL4) GetCrcString() string {
+	return "4588385f"
+}
+
+// PuntReason represents VPP binary API type 'punt_reason':
+type PuntReason struct {
+	ID          uint32
+	XXX_NameLen uint32 `struc:"sizeof=Name"`
+	Name        string
+}
+
+func (*PuntReason) GetTypeName() string {
+	return "punt_reason"
+}
+func (*PuntReason) GetCrcString() string {
+	return "fc7d34bc"
+}
+
+/* Unions */
+
+// AddressUnion represents VPP binary API union 'address_union':
+type AddressUnion struct {
+	Union_data [16]byte
+}
+
+func (*AddressUnion) GetTypeName() string {
+	return "address_union"
+}
+func (*AddressUnion) GetCrcString() string {
+	return "d68a2fb4"
+}
+
+func AddressUnionIP4(a IP4Address) (u AddressUnion) {
+	u.SetIP4(a)
+	return
+}
+func (u *AddressUnion) SetIP4(a IP4Address) {
+	var b = new(bytes.Buffer)
+	if err := struc.Pack(b, &a); err != nil {
+		return
+	}
+	copy(u.Union_data[:], b.Bytes())
+}
+func (u *AddressUnion) GetIP4() (a IP4Address) {
+	var b = bytes.NewReader(u.Union_data[:])
+	struc.Unpack(b, &a)
+	return
+}
+
+func AddressUnionIP6(a IP6Address) (u AddressUnion) {
+	u.SetIP6(a)
+	return
+}
+func (u *AddressUnion) SetIP6(a IP6Address) {
+	var b = new(bytes.Buffer)
+	if err := struc.Pack(b, &a); err != nil {
+		return
+	}
+	copy(u.Union_data[:], b.Bytes())
+}
+func (u *AddressUnion) GetIP6() (a IP6Address) {
+	var b = bytes.NewReader(u.Union_data[:])
+	struc.Unpack(b, &a)
+	return
+}
+
+// PuntUnion represents VPP binary API union 'punt_union':
+type PuntUnion struct {
+	Union_data [4]byte
```

Comment:
@ondrej-fabry 
This union is generated with incorrect length. 

VPP API definition:
```
typedef punt_exception
{
    u32 id;
};
typedef punt_l4
{
    vl_api_address_family_t af;
    vl_api_ip_proto_t protocol;
    u16 port;
};
union punt_union
{
    vl_api_punt_exception_t exception;
    vl_api_punt_l4_t l4;
};
```

Obviously, the `punt_l4` is longer (2x enum + u16 = 10 bytes), however it is generated only with 4. Changing it manually to 10 makes it work (well, kind of because of the broken UDP port resolution in the current VPP version).

Reply:
Looking into the issue with generating unions in GoVPP binapi-generator. This is first time I see union used for composed types.